File: backend/src/main/java/lt/ca/javau10/sakila/services/UserService.java
package lt.ca.javau10.sakila.services;

import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import jakarta.transaction.Transactional;
import lt.ca.javau10.sakila.exceptions.ResourceNotFoundException;
import lt.ca.javau10.sakila.models.Address;
import lt.ca.javau10.sakila.models.Customer;
import lt.ca.javau10.sakila.models.User;
import lt.ca.javau10.sakila.models.dto.PersonalInfoDto;
import lt.ca.javau10.sakila.repositories.AddressRepository;
import lt.ca.javau10.sakila.repositories.CustomerRepository;
import lt.ca.javau10.sakila.repositories.UserRepository;

@Service
public class UserService {

    private UserRepository userRepository;
    private CustomerRepository  customerRepository;
    private AddressRepository  addressRepository;
    private final PasswordEncoder passwordEncoder;
    
    public UserService(UserRepository userRepository, CustomerRepository  customerRepository, AddressRepository  addressRepository, PasswordEncoder passwordEncoder) {
		this.userRepository = userRepository;
		this.customerRepository = customerRepository;
		this.addressRepository = addressRepository;
		this.passwordEncoder = passwordEncoder;
	}
    
    
    public Double getUserBalance(String username) {
        User user = userRepository.findByUsername(username)
            .orElseThrow(() -> new UsernameNotFoundException("User not found"));
        return user.getBalance();
    }

    @Transactional
    public Double addBalance(String username, Double amount) {
        User user = userRepository.findByUsername(username)
            .orElseThrow(() -> new UsernameNotFoundException("User not found"));
        Double newBalance = user.getBalance() + amount;
        user.setBalance(newBalance);
        userRepository.save(user);
        return newBalance;
    }
    
  //Change password
    public void changePassword(String username, String currentPassword, String newPassword) {
        User user = userRepository.findByUsername(username)
            .orElseThrow(() -> new UsernameNotFoundException("User not found"));
        // Verify the current password
        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new BadCredentialsException("Current password is incorrect");
        }
        // Update the user's password
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }
    
    public int getUserIdByUsername(String username) {
        User user = userRepository.findByUsername(username)
            .orElseThrow(() -> new UsernameNotFoundException("User not found"));
        return user.getUserId();
    }
    
    //Get personal information based on user ID
    public PersonalInfoDto getPersonalInfo(int userId) {
        Customer customer = customerRepository.findByUser_UserId(userId)
            .orElseThrow(() -> new ResourceNotFoundException("Customer not found"));
        Address address = addressRepository.findById(customer.getAddress().getAddressId())
            .orElseThrow(() -> new ResourceNotFoundException("Address not found"));

        PersonalInfoDto dto = new PersonalInfoDto();
        dto.setFirstName(customer.getFirstName());
        dto.setLastName(customer.getLastName());
        dto.setEmail(customer.getEmail());
        dto.setPhone(address.getPhone());

        return dto;
    }

    //Update personal information
    public void updatePersonalInfo(int userId, PersonalInfoDto personalInfoDto) {
        Customer customer = customerRepository.findByUser_UserId(userId)
            .orElseThrow(() -> new ResourceNotFoundException("Customer not found"));
        Address address = addressRepository.findById(customer.getAddress().getAddressId())
            .orElseThrow(() -> new ResourceNotFoundException("Address not found"));

        customer.setFirstName(personalInfoDto.getFirstName());
        customer.setLastName(personalInfoDto.getLastName());
        customer.setEmail(personalInfoDto.getEmail());
        address.setPhone(personalInfoDto.getPhone());

        customerRepository.save(customer);
        addressRepository.save(address);
    }
}

File: frontend/src/services/userService.js
import createAxiosInstance from './axiosInstance';

const axiosInstance = createAxiosInstance('/api/user');

// BALANCE
// Fetch balance
export const getBalance = async () => {
    try {
        const response = await axiosInstance.get('/balance');
        return response.data;
    } catch (error) {
        console.error("Error fetching balance:", error);
        throw error;
    }
};

// Add balance
export const addBalance = async (amount) => {
    try {
        const response = await axiosInstance.post('/balance/add', { amount });
        return response.data;
    } catch (error) {
        console.error("Error adding balance:", error);
        throw error;
    }
};

// PROFILE
// Fetch personal information
export const getPersonalInfo = async () => {
    try {
        const response = await axiosInstance.get('/profile/personal-info');
        return response.data;
    } catch (error) {
        console.error("Error fetching personal information:", error);
        throw error;
    }
};

// Update personal information
export const updatePersonalInfo = async (personalInfo) => {
    try {
        const response = await axiosInstance.put('/profile/personal-info', personalInfo);
        return response.data;
    } catch (error) {
        console.error("Error updating personal information:", error);
        throw error;
    }
};

// Fetch address information
export const getAddressInfo = async () => {
    try {
        const response = await axiosInstance.get('/profile/address-info');
        return response.data;
    } catch (error) {
        console.error("Error fetching address information:", error);
        throw error;
    }
};

// Update address information
export const updateAddressInfo = async (addressInfo) => {
    try {
        const response = await axiosInstance.put('/profile/address-info', addressInfo);
        return response.data;
    } catch (error) {
        console.error("Error updating address information:", error);
        throw error;
    }
};

// SECURITY
// Change password
export const changePassword = async (currentPassword, newPassword) => {
    try {
        const response = await axiosInstance.post('/security/change-password', {
            currentPassword,
            newPassword
        });
        return response.data;
    } catch (error) {
        console.error("Error changing password:", error);
        throw error;
    }
};